blog/2022-01-01-unit-testing.mdx

```diff
@@ -0,0 +1,641 @@
+
+<AnchoredP id='first-encounter'>
+	🔸 When did you first encounter unit testing?
+</AnchoredP>
+
+🔹 It was a long time ago in a galaxy far, far away... I barely had a few years of experience.
+I've been working in the same company for most of my career, and compared to my day-to-day, I thought I had seen it all.
+I started looking online in guideliens, best practices, patterns and so on... I came across a lot of exciting and above all new topics.
+I seemed like I stepped in a new world that had nothing to do with my daily life in which I was stuck. The _testing_ was obviously one of those topics.
+However, I quickly realised that I had to progress in other subjets before I could introduce tests in my projects. It took me quite a long time before I was really fit to work in a test-driven approach.
+
+<AnchoredP id='what-is-it'>
+	🔸 Ok! However, just not to lose our readers... can you explain what a unit
+	test is please?
+</AnchoredP>
+
+🔹 Yes, of course! A unit test is a test that checks a _unit of code_. This concept of unit of code varies according to the testing schools.
+The London (or Mockist) School will see it as the smallest chunk of code so we are talking about a class or a method.
+On the other hand, the Detroit (or Classicist) School will see it as a group of classes or methods.
+For the most curious of you, here's an [article](https://medium.com/@adrianbooth/test-driven-development-wars-detroit-vs-london-classicist-vs-mockist-9956c78ae95f) which gives details on the differences between both schools. I want to clarify that no school is better than the other, each one having its pros and cons.
+It's only a matter of preference and compromise.
+The main difference is mainly the size of your _System Under Test_ (SUT) and the relationship between the different collaborators.
+
+But in the end, a unit test is simply **a piece of code validating that another piece of code has the expected result and/or side effect against a given scenario**.
+
+<AnchoredP id='hierarchy-position'>
+	🔸 Alright. Where is it in this "test hierarchy"?
+</AnchoredP>
+
+🔹 It looks like the Agile Testing Pyramid but from left to right instead of bottom to top.
+The more on the left, the smaller your scope is and the most numerous and faster your tests will be.
+If you read it the other way and apply a reverse logic, it works too.
+
+We are on the "unit" part for this topic since there is technically nothing smaller than a unit. A unit test should be executed on a standalone basis in a sandboxed environment.
+That is, a unit test has no impact on the outside of its scope, and if you run several of them in parallel, they must not have any side effect between them.
+
+This finally means that in a unit test: we do not contact a database, we avoid making HTTP requests, access file system, etc. In the same way, we are not going to contact the other dependencies of the solution.
```

Comment:
we avoid reaching a database, making HTTP requests, accessing the file system, etc.